### PR TITLE
fix: InfluxQL planner promotes integer division to floats

### LINKED ISSUE #8168: InfluxQL: Integer fields using division operator returns 0
__Steps to reproduce:__

Using `issue_6112` data set in IOx

1. Execute a raw query:

    ```sql
    SELECT bytes_free / bytes_used AS result FROM disk
    ```

    OR an aggregate query:

    ```sql
    SELECT SUM(bytes_free) / SUM(bytes_used) AS result FROM disk WHERE time >= '2022-10-31T02:00:00Z' AND time <= '2022-10-31T02:00:10Z' GROUP BY time(5s) FILL(null)
    ```

__Expected behaviour:__

Output of `result` column should be float for the raw query:

```text
name: disk
time                 result
----                 ------
2022-10-31T02:00:00Z 0.005613224283335911
2022-10-31T02:00:00Z 0.007702971146013462
2022-10-31T02:00:00Z 0.006984786047936768
2022-10-31T02:00:10Z 0.0077149723818756505
2022-10-31T02:00:10Z 0.007000528400759146
2022-10-31T02:00:10Z 0.005636096491427584
```

and the aggregate query:

```text
name: disk
time                 result
----                 ------
2022-10-31T02:00:00Z 0.006984786047936768
2022-10-31T02:00:05Z
2022-10-31T02:00:10Z 0.007000528400759146
```

__Actual behaviour:__

Output are `0`s for the raw query:

```text
name: disk
time                 result
----                 ------
2022-10-31T02:00:00Z 0
2022-10-31T02:00:00Z 0
2022-10-31T02:00:00Z 0
2022-10-31T02:00:10Z 0
2022-10-31T02:00:10Z 0
2022-10-31T02:00:10Z 0
```

and the aggregate query:

```text
name: disk
time                 result
----                 ------
2022-10-31T02:00:00Z 0
2022-10-31T02:00:05Z
2022-10-31T02:00:10Z 0
```

## InfluxQL OG

When evaluating a division operation, InfluxQL may set the `IntegerFloatDivision` flag to `true` to perform the division as floating-point values, [per this implementation](https://github.com/influxdata/influxql/blob/802555d6b3a35cd464a6d8afa2a6511002cf3c2c/ast.go#L4334-L4345):

```go
case DIV:
	if v.IntegerFloatDivision {
		if rhs == 0 {
			return float64(0)
		}
		return float64(lhs) / float64(rhs)
	}

	if rhs == 0 {
		return int64(0)
	}
	return lhs / rhs
```

And this flag is set to true by InfluxDB per [this source](https://github.com/influxdata/influxdb/blob/3372d3b878ebcba708dc9edfce7ea83cc8152393/query/cursor.go#L178).

> **Note**
>
> InfluxQL OG only promotes division operations to `float` when both operands are `integer`. If one of the operands is `unsigned`, the resulting expression is `unsigned`.

### PR DESCRIPTION
**Pull request by stuartcarnie**
Closes #8168

Teaches the InfluxQL planner to promote expressions from integer to float for the division operator.

### CODE REVIEW

File: influxdb_iox/tests/query_tests/cases/in/issue_6112.influxql.expected [hunk 1]
```diff
@@ -372,7 +372,7 @@ name: m0
...
 +---------------------+-----+-------+---------------------+--------------------+--------------------+--------------------+-------------------+-----------+----------------------+-----------+------------------+-------------------+--------------------+-------------------+-------+-------+-------+-------+
-| 2022-10-31T02:00:00 | 101 | 101.0 | 0.45202578717835057 | 0.8920048697881602 | 0.5067526002248183 | 1.5707963267948966 | 1.560895660206908 | 1.5509969 | 7.307059979368067e43 | 2.2194037 | 4.61512051684126 | 6.658211482751795 | 2.0043213737826426 | 10.04987562112089 | 10201 | 101.0 | 101.0 | 101.0 |
+| 2022-10-31T02:00:00 | 101 | 101.0 | 0.45202578717835057 | 0.8920048697881602 | 0.5067526002248183 | 1.5608951749237256 | 1.560895660206908 | 1.5509969 | 7.307059979368067e43 | 2.2194037 | 4.61512051684126 | 6.658211482751795 | 2.0043213737826426 | 10.04987562112089 | 10201 | 101.0 | 101.0 | 101.0 |
```
Comment (stuartcarnie) [inline influxdb_iox/tests/query_tests/cases/in/issue_6112.influxql.expected:375]: The expression `acos(1/i64)` now matches InfluxQL OG output, as the division causes the operands to be cast to floats.

File: iox_query_influxql/src/plan/expr_type_evaluator.rs [hunk 2]
```diff
@@ -93,13 +93,13 @@ impl<'a> TypeEvaluator<'a> {
...
-                || {
+            (Some(lhs), Some(rhs)) => {
+                Ok(Some(binary_data_type(lhs, op, rhs).ok_or_else(|| {
                     error::map::query(format!(
                         "incompatible operands for operator {op}: {lhs} and {rhs}"
                     ))
-                },
-            )?)),
+                })?))
+            }
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/expr_type_evaluator.rs:102]: Defies logic why replacing `expr.op` with the local variable `op` caused this to be reformatted 🤷🏻

File: iox_query_influxql/src/plan/expr_type_evaluator.rs [hunk 3]
```diff
@@ -399,6 +399,13 @@ fn binary_data_type(
         (Float, Add | Sub | Mul | Div | Mod, Float | Integer | Unsigned)
         | (Integer | Unsigned, Add | Sub | Mul | Div | Mod, Float) => Some(Float),
 
+        // Integers using the division operator are always float
+        //
+        // See:
+        // * https://github.com/influxdata/influxql/blob/802555d6b3a35cd464a6d8afa2a6511002cf3c2c/ast.go#L4335-L4340
+        // * https://github.com/influxdata/influxdb/blob/3372d3b878ebcba708dc9edfce7ea83cc8152393/query/cursor.go#L178
+        (Integer, Div, Integer) => Some(Float),
+
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/expr_type_evaluator.rs:408]: This wasn't necessary for the fix, but for completeness, we want the data type during this phase to be correctly determined.

File: iox_query_influxql/src/plan/planner_rewrite_expression.rs [hunk 4]
```diff
@@ -341,10 +341,11 @@ fn rewrite_expr(expr: Expr, schemas: &Schemas) -> Result<Expr> {
...
                         DataType::Int64
                     ) => match op {
-                        // Like Float64, dividing by zero should return 0 for InfluxQL
+                        // Like Float64, dividing by zero should return 0 for InfluxQL, and
+                        // the expression should be promoted to Float64, so cast both sides.
                         //
-                        // See: https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4338-L4340
-                        Operator::Divide => yes(coalesce(vec![expr, lit(0_i64)])),
+                        // See: https://github.com/influxdata/influxql/blob/1ba470371ec093d57a726b143fe6ccbacf1b452b/ast.go#L4331-L4336
+                        Operator::Divide => yes(coalesce(vec![binary_expr(cast((**left).clone(), DataType::Float64), Operator::Divide, cast((**right).clone(), DataType::Float64)), lit(0_f64)])),
```
Comment (stuartcarnie) [inline iox_query_influxql/src/plan/planner_rewrite_expression.rs:348]: This is the fix to match InfluxQL OG behaviour for which only promotes division expressions to float when both operands are integers. That is, the division operator does not cause unsigned operands or mixed integer and unsigned operands to be promoted to float. Strange.

**Review by mhilton — APPROVED** [review]
LGTM

**Pull request merged**